From b9638a1b76596c79e55fbfca28de3fd9a1685727 Mon Sep 17 00:00:00 2001
From: Sandor Yu <Sandor.yu@nxp.com>
Date: Thu, 27 Sep 2018 14:28:58 +0800
Subject: [PATCH 4753/5242] MLK-18895: hdmi: Reset HDMI PHY when
 enable/disable HDCP function

commit  cb3b73e4234a1c4d0fe117ee236967b447cf280e from
https://source.codeaurora.org/external/imx/linux-imx.git

Some HDMI Sink may failed to work when running HDCP enable/disable
stress test. HDMI sink may not support change HDCP state on the fly.
So reset HDMI PHY and controller to resolve the issue.

Signed-off-by: Sandor Yu <Sandor.yu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mxc_hdmi.c |    9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/video/fbdev/mxc/mxc_hdmi.c b/drivers/video/fbdev/mxc/mxc_hdmi.c
index 9e84480..c8aeadd 100644
--- a/drivers/video/fbdev/mxc/mxc_hdmi.c
+++ b/drivers/video/fbdev/mxc/mxc_hdmi.c
@@ -194,6 +194,7 @@ struct mxc_hdmi {
 extern void mxc_hdmi_cec_handle(u16 cec_stat);
 
 static void mxc_hdmi_setup(struct mxc_hdmi *hdmi, unsigned long event);
+static void mxc_hdmi_phy_disable(struct mxc_hdmi *hdmi);
 static void hdmi_enable_overflow_interrupts(void);
 static void hdmi_disable_overflow_interrupts(void);
 
@@ -337,6 +338,7 @@ static ssize_t mxc_hdmi_store_hdcp_enable(struct device *dev,
 	char event_string[32];
 	char *envp[] = { event_string, NULL };
 	unsigned long value;
+	u8 clkdis;
 	int ret;
 
 	ret = kstrtoul(buf, 10, &value);
@@ -345,6 +347,13 @@ static ssize_t mxc_hdmi_store_hdcp_enable(struct device *dev,
 
 	hdmi->hdmi_data.hdcp_enable = value;
 
+	/* Disable All HDMI clock and HDMI PHY */
+	clkdis = hdmi_readb(HDMI_MC_CLKDIS);
+	clkdis |= 0x5f;
+	hdmi_writeb(clkdis, HDMI_MC_CLKDIS);
+	mxc_hdmi_phy_disable(hdmi);
+	hdmi_disable_overflow_interrupts();
+
 	/* Reconfig HDMI for HDCP */
 	mxc_hdmi_setup(hdmi, 0);
 
-- 
1.7.9.5

